### ISSUE TITLE
Control queue and playback

### ISSUE THREAD

**Issue by CoDEmanX** [spotify/web-api#15]
There's nothing in the Web API to control a user's current queue and playback (including seek time), but Apps have been deprecated on the other hand and Spotify Connect is proprietary. That leaves no option to implement a new PartiQ\* - that's a shame.

Can we expect an open Spotify Connect or new Endpoints in the Web API to remote-control Spotify within like a year or should we forget that there was something as nice as PartiQ once?

I'd prefer additional API features over a built-in solution, because community might want to do queuing differently (e.g. let people vote on suggested tracks and play what receives the most votes instead of first-in/first-out, or group suggested tracks by genre and tempo to avoid too hard switches).

But it doesn't end there: one could create a web app that plays what a certain user plays (like a request show in the radio, but without the need to capture audio of the DJ and send it to all listeners). Maybe even play music on multiple wifi speakers in sync? Or a custom last.fm scrobbler...
- = Former Spotify app that let your party guests decide what music to play in a quite democratic fashion

_Edit: Added "including seek time"._ -@thelinmichael

**Comment by thelinmichael** [comment]
Thanks for your feedback @CoDEmanX. I'm going to add the enhancement label to this issue since it's a very valid feature request, but I can not make any promises that this feature will be available regardless of timespan and scope.

**Comment by markbrouch** [comment]
+1 I would love to see this implemented.

**Comment by Eastborn** [comment]
+1 this would be great

**Comment by xeno** [comment]
+1 I *neeeed* this please. :)

**Comment by sashaperigo** [comment]
+1 I would love access to a user's queue!

**Comment by Hagfjall** [comment]
+1 I would love this, to be able to control the playback for a user.

**Comment by auhsor** [comment]
Please add this support!

**Comment by mk83** [comment]
+1 And be quick about it.

**Comment by nicohvi** [comment]
+1

**Comment by hunterbridges** [comment]
+1

**Comment by petermakowski** [comment]
+1

**Comment by rafaelbandeira3** [comment]
I am working on party playlists for my wedding, and as a developer I would love to use my premium account on spotify just for streaming while I can have custom interfaces and options, like manually defined crossfade for different songs and more, all of which I can work on my own. So, +1

**Comment by alairock** [comment]
+1

**Comment by overnine** [comment]
+1

**Comment by kevinweber** [comment]
+1

**Comment by Clarkinator** [comment]
+1

**Comment by calvin-evans** [comment]
+1 for queueing tracks via the API

**Comment by mexitalian** [comment]
+1

**Comment by forstermatth** [comment]
+1 I'd love to use a feature like this to create a slack chat script to control music in the office

**Comment by Gavinator98** [comment]
+1

**Comment by gartnera** [comment]
+1

**Comment by malte-p12** [comment]
+1

**Comment by bash** [comment]
+1

**Comment by ericwebster** [comment]
+1

**Comment by JoshCheek** [comment]
Looks like you can get some reasonable functionality with [applescript](https://developer.spotify.com/applescript-api/) 

I played around and figured out how to do some of the things. Note that I do not know AppleScript, I just looked at the examples and [the docs](https://developer.apple.com/library/mac/documentation/AppleScript/Conceptual/AppleScriptLangGuide/introduction/ASLR_intro.html).

Choosing content
-----------------------

```
# A track
$ osascript -e 'tell app "Spotify" to play track "spotify:track:4KWntZBFeQOdoYNgPwcEGe"'

# A playlist
$ osascript -e 'tell app "Spotify" to play track "spotify:user:josh_cheek:playlist:3WfJpjKmq84WjORNHHsZMk"'
```

Playing / pausing
-----------------------

```
# Resume playback
$ osascript -e 'tell app "Spotify" to pause'

# Pause playback
$ osascript -e 'tell app "Spotify" to play'

# Toggle play/pause
$ osascript -e 'tell app "Spotify" to playpause'
```

Current Track
------------------

You can get some info about the current track like this (there's more info you can get, open the AppleScript Editor and go to File -> Open Dictionary -> Spotify -> Track)

```
$ osascript -e 'tell application "Spotify"
       (artist of current track as string) & "\n" & (album of current track as string) & "\n" & (name of current track) & "\n" & (spotify url of current track as string)
   end tell'
```

Next and Prev track
--------------------------

$ osascript -e 'tell application "Spotify" to next track'
$ osascript -e 'tell application "Spotify" to previous track'


Volume
----------

```
$ osascript -e 'tell application "Spotify"
     get sound volume
   end tell'

$ osascript -e 'tell application "Spotify"
     set sound volume to 50
   end tell'
```

Find out if its playing/paused/stopped
--------------------------------------------------

```
$ osascript -e 'tell application "Spotify"
     get player state
   end tell'
```

Position within the track
-------------------------------

```
$ osascript -e 'tell application "Spotify"
     get player position
   end tell'

$ osascript -e 'tell application "Spotify"
     set player position to 30
   end tell'
```

Shuffling
------------

Is it on / turn it on

```
$  osascript -e 'tell application "Spotify"
     get shuffling
   end tell'

$  osascript -e 'tell application "Spotify"
     set shuffling to true
   end tell'
```

Example
------------

[Here](https://gist.github.com/JoshCheek/63f2f0edbb4f429e003438a41caa0142)'s a remote control script I wrote in Ruby, using most/all of these things. You run it and press p, q, h, H, j, k, l, L, -, + to command the player (map is at the bottom of the file)

**Comment by alairock** [comment]
@JoshCheek This is excellent work! 

I want to also put out there, that this "bandaid" isn't a sufficient enough solution to mark this as `resolved`. Native controls through an official REST-API is really the solution most of us would need to give the flexibility we are looking for.

**Comment by mateusmaso** [comment]
+1

**Comment by overnine** [comment]
@JoshCheek Still, play queue control is one of the major issues. Doesn't look like your script can control that?

**Comment by psrivast** [comment]
+1

**Comment by JoshCheek** [comment]
@overnine it uses the AppleScript API, so if the app is on a Mac, and its the same computer as the program that is trying to control it, then it can, otherwise it cannot. Obviously this should be readded to the API, but for now, its enough to build a jukebox for my stereo, I'll put a frontend on it and allow my students to hit it to change songs and things.

**Comment by matthewdias** [comment]
This issue more refers to the control of a user's queue, eg. adding songs to queue, reordering, etc. This isn't possible right now with AppleScript.

**Comment by oskaremilsson** [comment]
+1

**Comment by jws121295** [comment]
+1

**Comment by LazyTarget** [comment]
+1

**Comment by crwgregory** [comment]
+1!

**Comment by jordanranson** [comment]
I'd love to be able to add to my queue via the API or AppleScript. Rdio had these features and unfortunately they shut down and I had to swap to Spotify which is missing them...

**Comment by Hockeyfan360** [comment]
This is something I'm looking to use in an application I'm going to be building. +1

**Comment by ionTea** [comment]
+1

**Comment by jordanranson** [comment]
It's been months since an update from the devs. Any word on when this will be implemented?

**Comment by admgre** [comment]
+1

**Comment by greenrd** [comment]
Please stop commenting +1 - it leaves a notification for everyone else - please just use the thumbs up button under the original post.

**Comment by spikespaz** [comment]
+1

**Comment by TigerC10** [comment]
+1

**Comment by FleshFear** [comment]
This could brilliant to control my gramafon, using spotify connect through my home automation app.

**Comment by udany** [comment]
Really could use this... +1

**Comment by goodwince** [comment]
Following! +1

**Comment by jjabba** [comment]
+1

**Comment by greeny** [comment]
:+1:

**Comment by quickskape** [comment]
👍

**Comment by greenrd** [comment]
One again, please stop posting thumbs-up posts - please just click the thumbs-up button on the original post.

**Comment by TigerC10** [comment]
> The squeaky wheel gets the grease.

Compared to the number of +1s and thumbs, there's such a small minority of you that _don't_ want to get these +1 notifications.  There's only what, 4 of you?  5 of you?  Out of the roughly 50 people that have expressed themselves in this thread?

My inner troll wants to `moar notifications!` and `+1` the heck out of this issue.  But that would be immature.  But in all seriousness, if you don't want to get them then there's the unsubscribe button just to the right.  Stop complaining about notifications, you have the power to turn them off.

I, for one, invite everyone to `+1`.  Go ahead!  Spam me, and by extension Spotify, with notifications for how much you want this feature.  Because only by making noise will we get their attention to prioritize it.

**Comment by BjoernPetersen** [comment]
There are only a few people who said "stop commenting +1" because they it's already been said and that should be enough on GitHub. The devs are aware of this issue and further notifications are just annoying, not at all helpful.

I subscribed to this issue thread to stay up-to-date with relevant updates on the issue, not to get "+1"-notifications. 

For everyone wanting to comment "+1" in the future: please click the thumbs-up button for this issue and then maybe go ahead and follow TigerC10's invitation to spam him, but please stop commenting with "+1"-like comments here.

**Comment by TigerC10** [comment]
I disagree, keep the issue active with `+1` comments. 😃 If they are so annoying, the devs may be incentivized to address the issue to make it stop.  That's how it is helpful.

For everyone wanting to comment `+1` in the future: do it! :trollface:

**Comment by udany** [comment]
Yeah, do please +1, it's such a "basic feature" for enabling interesting
integrations it's almost absurd it hasn't been prioritized yet

**Comment by RBdevelop** [comment]
+1  aka PLUS ONE

**Comment by hunterbridges** [comment]
-1... I'm gonna make y'all really work for it

**Comment by JoshCheek** [comment]
+1

**Comment by 1youri** [comment]
+1

**Comment by greeny** [comment]
Just a fallback solution for some of the ideas I have seen here:

Create a playlist, get its identifier and add songs to that playlist instead. Let the device play from that playlist. Added songs to playlist would be this way added to "currently playing queue", which is basically whole playlist. (dont forget to turn off shuffle mode)

This does not solve the issue we are talking about, but it is at least some solution for us non-Applers, who cannot use that AppleScript thingy.

**Comment by udany** [comment]
I thought about doing this but wasn't sure how it would behave if the
playlist changed after it started playing it, good to know that works.

But it's still not a solution to what I need it to do which would be a
voting system, I'd need to be able to skip the current song and set the
next one just before the current one ends to give it as much voting time as
possible...

I'm considering creating a user script to work with the web player, that
might work.

**Comment by TigerC10** [comment]
The other fallback that I was working on was leveraging https://github.com/jonas-lundqvist/vlc-spotify to queue up songs in VLC.  The only problem with that was that you have to get a special app key from spotify to be able to get libspotify to work.  Unfortunately, spotify doesn't issue such keys anymore (though, if you're clever enough, you can still find the page to generate one).

However, after getting libspotify to compile I'm still having trouble getting vlc-spotify to compile on Mac.  I could probably get it to compile on Linux just fine, maybe I should set up a VM and see if I can get it working the way I expect.

**Comment by jscholes** [comment]
@udany
> I'd need to be able to skip the current song and set the next one just before the current one ends to give it as much voting time as possible...

This is pretty much what Last.fm do with their Spotify Remote Control Bridge integration.  There is no way, even with that control interface, to queue up a song, so instead they listen for an event to be fired that says a song has just finished, then send a play command with the next URI on their list.  A bit hacky, but it generally works.  You could also constantly monitor the current playback time and trigger your own queue advancement when a song is about to finish.

I won't go into any detail about the interface, as it's unsupported and not publically documented (by Spotify at least).  To my knowledge the only way to gain access to the official docs is to become a Spotify partner.  But if you're going to start hacking around with user scripts, you might want to give this a look instead.  You could easily create a headless app and UI to do what you're after.

**Comment by crwgregory** [comment]
Adding songs to a playlist while playing the playlist does not work. I tried it. The desktop app does not live refresh when a song gets added to the playlist using the web API, it has to go through a process of 'rebooting' to then retrieve it's new songs.

I created a *functioning* Spotify Music app that anyone could connect to on the local network and add songs to a playlist. The songs on the playlist could be voted on by any user and the number of votes a song would have would determine it's position on the playlist.

The hardest part of getting this app to work was exactly what this issue represents. I tried compiling v8 libspotify libraries to use during runtime and control the client. However libspotify is so out of date that any library built on top of it only compiles with node 0.12, so good luck getting a modern node server to run it. 

The hack that I implemented was making spotify believe that I was it's browser brother and making requests to play (the same way open.spotify works). The spotify desktop client runs a server on one of the 4*** ports that you can utilize. I would save the queue of songs in my app, then get the duration of the song I was about to play, then simply set a timeout for that duration. Once that was up I would just make a request with the next song in my playlist. 

My app was 100% functional and was used everyday at my company. There is a library out there that I used to make this happen (the requests to the spotify client). I'm not going to link it here for obvious reasons. But if your clever you can find it.

**Comment by Neo-Desktop** [comment]
Ticket #272 was closed out as a duplicate of this ticket

@CoDEmanX would it be possible to add "controlling the seek time" to this request?

**Comment by thelinmichael** [comment]
@Neo-Desktop, added seek time to the original post.

**Comment by RWOverdijk** [comment]
👍

**Comment by jordanranson** [comment]
The lack of audio control/playback is keeping me from reviving my app I made for Rdio. A shame they had to go out of business. Any update on this?

**Comment by lilalinux** [comment]
+1
It's unbelievable that the most important function for a music streaming service "playing music" isn't supported by the API.

**Comment by brandoncornel** [comment]
+1

**Comment by pinetrees** [comment]
+1

**Comment by AnderssonPeter** [comment]
@thelinmichael are there any news on this?
Im currently on a Windows 10 Mobile and the Spotify app there does not support controlling other devices if you where to add this functionality we could make our own Spotify remote app.

**Comment by drsounds** [comment]
I am afraid there is a specific reason why we haven't seen any playback API for the desktop and that is that people have made apps such Eternify, where artists could crowdsource streams in order to get more money. I think it's because this Spotify are not providing playback control interface to the public. On mobile the risk is much less that is happen, since mobiles can't be used as a server. 
 Playback control api on desktop could be abused for a lot of reasons, which could be costly for Spotify and I think it's the reason Spotify has chosen to not provide that functionality to the public.

**Comment by DaweedM** [comment]
+1

**Comment by DeckardKane** [comment]
@drsounds That sounds like a plausible reason for Spotify to not implement this. Could they just confirm it in a statement saying "We are not implementing this feature and [x reason] is why not". I have been tearing my hair out looking for this functionality, but if it is confirmed to not be a planned feature, I will find another service to use. Just tell us please!

**Comment by TigerC10** [comment]
@DeckardKane / @drsounds - There's at least partial playback control since you can open up a track/playlist/album while one's already going.  So, even if the argument is that playback control can be abused - well so what?  It can already be abused.

Since libspotify isn't a thing anymore, it'd be really great to at least get this finished out.

**Comment by drsounds** [comment]
Spotify is bleeding financially so I assume they have to cut down on some
stuff because they just borrowed one million. I assume they must resolve
some problems before they can open it up

Den 22 feb. 2017 5:31 em skrev "Ian Cervantez" <notifications@github.com>:

**Comment by brongracz** [comment]
They dont understand that they are missing a market like this. That is automation. Siri can play music. Google can play music. If spotify added that playback control API, it would become automaticly a default playback service for smart homes etc. from brands other than apple and google. Also the spotify connect with playback api could help smartify various devices to make part of automated ecosystems. Making spotify connect a sort of a standard for audio streaming in smart homes etc. equivalent to zigbee ZLL protocol for lightning control.
"You want your smart home to talk to all audio devices, choose a device with spotify connect support" is for now not the case. In fact this could make me look for another subscription.

**Comment by drsounds** [comment]
They have a Spotify Connect API for that, but only for approved partners.
And I think the obstacle for not releasing it for everyone is called
Eternify - the app some dude created which could cheat Spotify streams so
artists could generate high royalties. I am afraid this concern is
preventing Spotify for expanding the Web API and release a new desktop API
that supports playback control.

Best regards,
Alexander Forselius (Dr. Sounds)
Musician, and programmer.
http://dr-sounds.com

Github <http://www.github.com/krikelin>
Twitter <http://www.twitter.com/drsounds>
Spotify <http://open.spotify.com/artist/2FOROU2Fdxew72QmueWSUy>
<http://www.youtube.com/watch?v=pjdujamrcds>
LinkedIn <http://se.linkedin.com/pub/alexander-forselius/13/b11/500/>


2017-03-05 22:57 GMT+01:00 brongracz <notifications@github.com>:

**Comment by kfrz** [comment]
Hey guys! I want to chime in because I've been using Spotify Premium for much longer than I've been involved in software development, but I've always wanted to write an extension to shuffle by album, so I can pick fifteen albums or playlists and shuffle. Lo and I see, it's not really possible. 

If anyone has suggestions on how it may be possible, I'm still interested in writing this, I just don't think it'll work with the current Web API.

**Comment by iangilman** [comment]
@kfrz I know this is a poor substitute for an actual playback API, but the workaround I've been using  for my app (http://www.letsfathom.com/) is to have my app make a new playlist and fill it with whatever I want (e.g. 15 albums in random order). I then listen to that playlist once and delete it.

Like I said, it's not as great as an actual playback API, but when you start thinking in terms of these single-use playlists, there's actually a lot of interesting things you can do.

**Comment by jscholes** [comment]
> I've always wanted to write an extension to shuffle by album, so I can pick fifteen albums or playlists and shuffle but play the whole record before shuffling to a different record.

The availability of an API to control Spotify's playback queue might indeed help with this, but it's not the only way.  You just have to think a little bit outside the box, and watch out for rate limits.  For example, you could:

1. Create a new playlist via the Web API.
2. Create a hash table/dict/whatever your language of choice refers to it as, use it to store your album/playlist objects.
3. For each album/playlist, fetch the list of tracks using the Web API and store them in an array which will be a value in your dictionary - this is the part which might bite you if you don't take care to avoid Spotify's rate limiting.  Once this bit is done, you should have a hash table with Spotify album/playlist URIs for the keys, and lists of matching tracks for the values.
4. Once you have an ordered list of tracks for each album/playlist, shuffle the dictionary, rather than the track lists.  This will give you the albums/playlists in a random order, but the tracks of each won't be shuffled.
5. Flatten the whole lot to a list of tracks and add them all to the playlist you created in step 1.

**Comment by kfrz** [comment]
@jscholes -- thanks, that was a great response. I figured I could do something like build a playlist on the fly, so that's probably the way I'll approach it.

**Comment by asmitter** [comment]
Thanks everyone for the patience, input and workarounds on these ideas.
We're very pleased to release the following set of endpoints for interacting with playback via Connect: https://developer.spotify.com/web-api/web-api-connect-endpoint-reference/

As they align pretty closely with the discussion here, I'll close this issue. This thread did evolve though, so please open new issues if these new endpoints don't do the trick.

**Comment by dreszczyk** [comment]
Is there a library what covers the [last changes](https://github.com/spotify/web-api/issues/15#issuecomment-290762329)?

**Comment by lodev09** [comment]
@dreszczyk here's my sbotify bot using the new playback stuff :)
https://github.com/lodev09/sbotify

**Comment by tjhrulz** [comment]
@asmitter Any chance we chould get a playpause command added to it?
